### ISSUE TITLE
Remote package management

### ISSUE THREAD

**Issue by uael** [xmake-io/xmake#69]
> A package manager or package management system is a collection of software tools that automates the process of installing, upgrading, configuring, and removing computer programs for a computer's operating system in a consistent manner.

> A package manager deals with packages, distributions of software and data in archive files. Packages contain metadata, such as the software's name, description of its purpose, version number, vendor, checksum, and a list of dependencies necessary for the software to run properly. Upon installation, metadata is stored in a local package database. Package managers typically maintain a database of software dependencies and version information to prevent software mismatches and missing prerequisites. They work closely with software repositories, binary repository managers, and app stores.

> Source: [package manager](https://en.wikipedia.org/wiki/Package_manager)

Remote package management roadmap:
- [x] detect tools and environment and wrap them
  - [x] git
  - [x] curl
  - [x] wget
  - [x] unzip/tar/7z/gzip
  - [x] ping (test speed)
- [x] Extractor wrapper
  - [x] tar 
  - [x] unzip
  - [x] gzip
  - [x] 7z
- [x] Cache management
- [x] semver
  - [x] parse
  - [x] satisfies
  - [x] select
- [ ] Multi language
  - [x] asm/c/c++
  - [x] objc/c++
  - [x] swift
  - [x] rust
  - [x] dlang
- [x] Multi platform
  - [x] linux
  - [x] macOS
  - [x] windows 
    - [x] git
    - [x] downloader (http[s], ftp ..)
    - [x] extractor (zip, tar.gz, 7z .. )
- [x] Command line tools
  - [x] require task
    - [x] install
      - [x] parse require info
      - [x] parse and load packages from repositories
      - [x] download package
        - [x] optimization: select the fastest url if exists multiple urls
      - [x] extract package
      - [x] build package 
        - [x] xmake.lua
          - [x] linux/macos
          - [x] windows
        - [x] makefile
          - [x] linux/macos
          - [x] windows
        - [x] configure/make
          - [x] linux/macos
        - [x] cmakelists.txt
          - [x] linux/macos
          - [x] windows (in progress ..)
        - [x] *.vcproj
          - [x] windows
      - [x] install package
      - [x] install basic dependent tools (git, curl, unzip ...)
    - [x] search
    - [x] clear
    - [x] remove
    - [x] list
  - [x] repo task
    - [x] add repository
    - [x] remove repository
    - [x] list all repositories
    - [x] clear all repositories and cache
    - [x] global or local repositories
- [x] xmake lua api
  - [x] `add_requires`: add package require info
  - [x] `add_repositories`: add repository url
  - [x] `add_packages`: add package dependencies to target
  - [x] `package`: package definition
    - [x] `set_urls`
    - [x] `add_versions`
    - [x] `set_homepage`
    - [x] `set_description`
    - [x] `add_deps`
    - [x] `[on|before|after]_load`: custom load script
    - [x] `[on|before|after]_install`: custom install script
    - [x] `[on|before|after]_test`: custom test script
    - [x] `add_patches`
    - [x] `add_links`
    - [x] `add_defines`
    - [x] `add_includedirs`

**Comment by uael** [comment]
References :
- [yarn](https://yarnpkg.com/en/docs/usage)
- [composer](https://getcomposer.org/doc/00-intro.md)
- [...](https://en.wikipedia.org/wiki/List_of_software_package_management_systems)

There is many other but they are good one

**Comment by waruqi** [comment]
Great! I have finished the `xmake repo` plugin task to manage repositories.

for example:

```bash
$ xmake repo --add origin https://github.com/tboox/xmake-repo.git
$ xmake repo --add other-repo https://github.com/xxxx/other-repo.git
```

Or use global `add_repositories` api in xmake.lua:

```lua
add_repositories("origin https://github.com/tboox/xmake-repo.git")
add_repositories("other-repo https://github.com/other/other-repo.git")
```

And the `xmake-repo` is the default repository, we need not add it manually.

**Comment by waruqi** [comment]
And we can add requires and add package dependencies (developing ..)

Add requires:

```lua
     add_requires("tboox.tbox >=1.5.1", "zlib >=1.2.11")
     add_requires("zlib master")
     add_requires("xmake-repo/tboox.tbox >=1.5.1")
     add_requires("https://github.com/tboox/tbox.git@tboox.tbox >=1.5.1")
```

Add package dependencies:

```lua
    target("test")
         add_packages("tboox.tbox", "zlib")
```

If you run `xmake deps` or `xmake deps --install`, xmake will install and update all required packages to local project repository directory.

**Comment by uael** [comment]
It's all great 👍  
I'm thinking about a way to avoid the need of `xxx-repo.git`. 
Without hosting, git remain the strongest alternative, there is still some other ways:
- `xxx.git` contains a 'provided packages' like `dist`, `bin` or something, which contains `xxx-repo.git/packages` content like `bin/<version>/<arch>/..`
- Provide an lua api to require an unknown repository by adding some rules like:
```lua
-- Github release tags
require("https://github.com/tboox/tbox.git >=1.5.1")
    -- github tags, checkout, ..
    set_kind("github tags")

    -- override the distribution file extension
    set_dist_extension("pkg.tar.gz")

    -- some other rules/kind/options
   (...)
```
Which is ugly, the idea need some rework..

**Comment by uael** [comment]
💡 If `https://github.com/xxxx/xxx-repo.git` represent the organisation packages, it could be automated by oauth

**Comment by uael** [comment]
How did you plan to handle that case?
```lua
add_requires("https://github.com/tboox/tbox.git@tboox.tbox >=1.5.1")
```

**Comment by TitanSnow** [comment]
Excuse me, I'm wondering what's the advantage of `xmake repo` that `git submodule` doesn't have.

**Comment by waruqi** [comment]
@uael I just place xmake.lua and other some text files (.e.g *.patch) in `xmake-repo` and I did not intend to place the binary file in this repository. 
And you can create yourself private repository to place some packages and binary files for each archs and platforms.So it will not be too large.

- `add_requires("https://github.com/tboox/tbox.git@tboox.tbox >=1.5.1")`

`add_requires` can process it directly If this project has complete tags like 1.2.1 or v1.x.x and it will define a package named `tboox.tbox`. We can use `add_packages("tboox.tbox")` to add package denpence to target.

If this git project has not tags, we can set a branch name, for example: `add_requires("https://github.com/tboox/tbox.git@tboox.tbox master")`

Or you can directly define `package("tboox.tbox")` to describe custem rules in `xmake.lua` if cannot get git tags, for example:

```lua
package("tboox.tbox")
      add_versions("1.5.1", "1.6.0", "1.6.1") -- we can add a version list if no git tags
      set_url("https://github.com/tboox/tbox/archive/v$(v1).$(v2).$(v3).tar.gz")
      on_build(function ()
           -- if need
      end)
      on_package(function ()
           -- if need
      end)

add_requires("tboox.tbox >=1.5.1")
```

**Comment by waruqi** [comment]
@TitanSnow `git submodule` is great, but `xmake repo` and `xmake deps` will provide more convenient cross-platform package management. 

Some advantages:
- Do not need the user to manage each third party package
- Better version control
- Support `.tar.gz` or `.zip` package if the dependent package does not have a git repo, but `git submodule` need it.
- Automated package compilation and linking to the given target.
- Support multiple platforms, architectures and languanges automaticlly.

In most cases, you only need to add two lines in `xmake.lua` to link dependent package automatically which exists in `xmake-repo`.

```lua
add_requires("pcre2 >=1.2.0", "zlib >= 1.2.11")
target("test")
    add_packages("pcre2", "zlib")
```

**Comment by waruqi** [comment]
@uael `semver.lua` is great 👍. I will consider using it.

**Comment by TitanSnow** [comment]
Sounds like fun. Like it:heart:. May it be hard work?

**Comment by uael** [comment]
@TitanSnow actually the following isn't possible with git submodule:
`foo` depends on `bar`;
`foo` depends on `baz`
`bar` depends on `baz`
then with package managment the `foo` vendor dir look like:
```
vendor
|__ bar
   |__ ...
|__ baz
   |__ ...
```

**Comment by uael** [comment]
In the case on c static library i also want to avoid compile time, is there a way to specify architecture binary url ? Or add an url kind with a downloadable .tar.gz containing all binaries for all architecture ? This file can for example follow .pkg structure.

**Comment by TitanSnow** [comment]
I got it @uael

**Comment by TitanSnow** [comment]
As for me I prefer "code once, compile everywhere" @uael

**Comment by uael** [comment]
"code once, compile everywhere" does not prevent to avoid compile time on large build with a bunch of dependencies :+1:

**Comment by uael** [comment]
@waruqi `repositories` shouldn't being cloned globally ? Like on on $(home)/.cache/xmake/repositories/
and use there like aptitude use ppa

**Comment by TitanSnow** [comment]
Seems right

**Comment by TitanSnow** [comment]
May *git large file storage* is suitable to you @uael

**Comment by waruqi** [comment]
@uael Yes, I'm planning to clone repositories globally in the first time and then only need to `git pull` to update them when the repositories need be updated.

**Comment by waruqi** [comment]
I added a `optional` mode for `add_requires("zlib >=1.2.11 optional")` api. If this package is optional dependent, we will do not download and install it automatically by default, unless the user runs `xmake deps --install zlib` to install it.

For example:

```lua
-- define package
package("mbedtls")
    set_url("git@github.com:ARMmbed/mbedtls.git")
    on_build(function (package)
    end)
    on_install(function (package)
    end)
package_end()

-- requires
add_requires("zlib >=1.2.11")
add_requires("mbedtls master optional")
add_requires("xmake-repo@tboox.tbox >=1.5.1")
add_requires("git@github.com:glennrp/libpng.git@libpng >=1.6.28")
```

**Comment by uael** [comment]
Love it :+1: 
Can be really useful for multiple and optional backend for example, is there a way to test if installed with the lua api ?

**Comment by waruqi** [comment]
You need test installed packages? xmake will first detect whether a package has been installe, then install it if it does not exist.

**Comment by uael** [comment]
More I look at your commits on repo, more I seriously envisage to use lua 👍 
Forgot to answer you! No I mean add some defines, for example, to know in the code if an optional package is installed or not.

**Comment by waruqi** [comment]
you can use `is_option("xxx")`, for example:

```lua
-- define option switch for `xmake config --test1=y|n`
option("test1")
     set_default(false)
     ..

-- require package for `xmake config --test2=y|n` or checking and installing automatically
add_requires("test2 >1.2")

target("xxxx")
    set_kind("binary")
    add_files("*.c")
    
    if is_option("test1") then
         add_defines("-DTEST1")
    end

    if is_option("test2") then
         add_defines("-DTEST2")
    end
```

`is_option` will return true If the custom option or package is enabled or installed.

**Comment by uael** [comment]
#75

**Comment by waruqi** [comment]
I rename `xmake deps` action to `xmake require`. for example

```bash
# install all package dependencies in xmake.lua
$ xmake require

# install the given required packages manually
$ xmake require zlib tboox.tbox
$ xmake require "zlib >=1.2.11" "tboox.tbox master"
$ xmake require "git@github.com:tboox/tbox.git@tboox.tbox >=1.6.0 <1.6.1 || master"
```

**Comment by TitanSnow** [comment]
By the way, may I ask how soon `repo` would be merged back to `master`? Or never? @waruqi

**Comment by waruqi** [comment]
@TitanSnow I will merge it into dev/master when I complete remote package management on linux or macOS.  Then I will continue to support it on windows.

May also need to wait for some time, but not too long.

**Comment by uael** [comment]
That's the prettier roadmap I never seen :1st_place_medal:

**Comment by TitanSnow** [comment]
Will phony package, aka *group*, be supported?

For example, if I wanna install development tools such as `gcc`, `make`, etc, all I need is to install `build-essential`
```console
$ sudo apt-get install build-essential
```

**Comment by waruqi** [comment]
You can add phony package without urls. for example:

```lua
package("devel-essential")
    add_requires("zlib >=1.2.11")
    add_requires("mbedtls master optional")
```

Then we can use `$xmake require devel-essential` to install them or add require in `xmake.lua`.

```lua
add_requires("devel-essential")
```

And run `$xmake require`

**Comment by TitanSnow** [comment]
Great:+1:

**Comment by TitanSnow** [comment]
```console
$ xmake require zlib            
updating repository ..
error: Cloning into '/home/ts/.xmake/repositories/xmake-repo'...
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

**?**

I think this is because of git protocol

```console
$ xmake repo --list
local repositories:

global repositories:
    xmake-repo git@github.com:tboox/xmake-repo.git

1 repositories were found!
```

Will https be better?

```console
$ xmake repo --list
local repositories:

global repositories:
    xmake-repo https://github.com/tboox/xmake-repo.git    ###

1 repositories were found!
```

**Comment by waruqi** [comment]
@TitanSnow  ok, I will fix it. 

Thanks! 😸

**Comment by TitanSnow** [comment]
```console
$ xmake repo --list
local repositories:

global repositories:
    xmake-repo https://github.com/tboox/xmake-repo.git
    ts-repo https://github.com/TitanSnow/xmake-repo.git

2 repositories were found!
$ xmake require 'TitanSnow.winpty >=0.4.2-xmake'
updating repository ..
  => downloading https://github.com/TitanSnow/winpty/archive/0.2.0.tar.gz ..
```

I specified `>=0.4.2-xmake`, why `0.2.0`?

package [TitanSnow.winpty](https://github.com/TitanSnow/xmake-repo/blob/master/packages/TitanSnow/winpty/xmake.lua)

**Comment by waruqi** [comment]
@TitanSnow I haven't implemented the semantic version yet and [semver.lua](https://github.com/tboox/xmake/blob/repo/xmake/core/base/semver.lua) only is stub. 

@uael is helping me to implement this module.  You can see [#75](https://github.com/tboox/xmake/pull/75)

**Comment by TitanSnow** [comment]
:v:

**Comment by uael** [comment]
Perfect win support :clap:

**Comment by uael** [comment]
About `semver` range parsing with lua pattern matching seems almost impossible or too much hardcoded, so, i'm currently adapting/writing a c version to wrap into the lua sandbox. Did you have some contraindication about adding an another :copyright: to xmake ? I can reinvent  the wheel in public domain otherwise

**Comment by uael** [comment]
I just realize that you get blocked by semver quickly

**Comment by waruqi** [comment]
@uael xmake need a lightweight semver, you can writing a c version, but try not to rely on other libraries.

What license do you want to add? As long as it is compatible with apache-2.0.

**Comment by uael** [comment]
I'm looking for some solutions, a good one is under 3-clause BSD which is compatible with apache-2.0, but I'm not an expert
I've to go fast as your are close to be blocked by this task !

**Comment by waruqi** [comment]
@uael No hurry. Is it this [semver-c](https://github.com/deoxxa/semver-c)? It seems a very rough clone of `node-semver`, but you can try it first.

I see the implementation of `node-semver`, it seems not very complicated if uses regex pattern.

If you need, I can help you to add regex match module into xmake.

**Comment by uael** [comment]
@waruqi I written it in c :smile: [uael/sv](https://github.com/uael/sv)
same grammar as the npm one need some more test and an xmake build but it's lightweight, fast and public domain :+1:

 :tada:  :tada: 500+ star and 100+ fork, BRAVO :wink:

**Comment by TitanSnow** [comment]
:tada: BRAVO

**Comment by waruqi** [comment]
@uael 
> @waruqi I written it in c 😄 uael/sv
same grammar as the npm one need some more test and an xmake build but it's lightweight, fast and public domain 👍

It is great! 👍

**Comment by waruqi** [comment]
@TitanSnow @uael Thanks! 😄

**Comment by uael** [comment]
@waruqi can you help me to pkg the thing into xmake ?

**Comment by waruqi** [comment]
@uael Ok, Can I put the sv code in the `xmake/core/src/sv` to compile together?

**Comment by uael** [comment]
Let me finish some stuff and test the all thing and go

**Comment by waruqi** [comment]
@uael ok. 👍

**Comment by waruqi** [comment]
@uael I have added libsv to xmake (`repo`). bb32a20aff39821e572583da16e00aa06c2a6309

**Comment by uael** [comment]
@waruqi Thank you :tada: 
How to wrap it in lua ?

**Comment by waruqi** [comment]
@uael You can refer to [core/src/xmake/readline/](https://github.com/tboox/xmake/tree/master/core/src/xmake/readline) and [readline-patch](https://github.com/tboox/xmake/commit/966876f494902efecb1ea475bade16cd96fe160c). for example:

```bash
core/src/xmake/semver
     --  some interface files 
     - xxx.c
     - xxx2.c
```

Then,  register them in `machine.c` and add `semver/*.c` in `core/src/xmake/makefile`

**Comment by TitanSnow** [comment]
QAQ

@waruqi A question: `return 0` means no return value, right?

**Comment by waruqi** [comment]
@TitanSnow Right, `return lua_api_return_value_count`

**Comment by TitanSnow** [comment]
Got it. I have never learned lua and did it just by reading your code and header of lua :smile:

**Comment by waruqi** [comment]
@uael @TitanSnow When I was doing `repo`, I found that xmake's detection function was really rough, it was not able to meet the needs of `repo`, so I intend to improve this feature first.

see [#83](https://github.com/tboox/xmake/issues/83)

**Comment by uael** [comment]
#98

**Comment by uael** [comment]
@waruqi 
I no more push all the semver table on select, the result is always `version, source` and here is the result after sorting and priority
```bash
uael@aero:~/project/xmake$ xmake lua
> semver = import('core.base.semver')
> semver.select(">2.0.1", {"2.0.0", "2.0.1", "2.0.2"}, {"v2.0.0", "v2.0.1", "v2.0.2", "v2.0.3", "v2.1.0-beta1", "v2.1.0-beta2"}, {"master", "v2.0", "v2.1"})
2.0.2 versions
> semver.select(">2.0.2", {"2.0.0", "2.0.1", "2.0.2"}, {"v2.0.0", "v2.0.1", "v2.0.2", "v2.0.3", "v2.1.0-beta1", "v2.1.0-beta2"}, {"master", "v2.0", "v2.1"})
v2.1.0-beta2 tags
> semver.select("^2", {"2.0.0", "2.0.1", "2.0.2"}, {"v2.0.0", "v2.0.1", "v2.0.2", "v2.0.3", "v2.1.0-beta1", "v2.1.0-beta2"}, {"master", "v2.0", "v2.1"})
2.0.2 versions
> semver.select("~2.0.9", {"2.0.0", "2.0.1", "2.0.2"}, {"v2.0.0", "v2.0.1", "v2.0.2", "v2.0.3", "v2.1.0-beta1", "v2.1.0-beta2"}, {"master", "v2.0", "v2.1"})
v2.1.0-beta2 tags
> semver.select("^2.0.9", {"2.0.0", "2.0.1", "2.0.2"}, {"v2.0.0", "v2.0.1", "v2.0.2", "v2.0.3", "v2.1.0-beta1", "v2.1.0-beta2"}, {"master", "v2.0", "v2.1"})
v2.1.0-beta2 tags
> semver.select("master", {"2.0.0", "2.0.1", "2.0.2"}, {"v2.0.0", "v2.0.1", "v2.0.2", "v2.0.3", "v2.1.0-beta1", "v2.1.0-beta2"}, {"master", "v2.0", "v2.1"})
master branches
> semver.select("v2.1", {"2.0.0", "2.0.1", "2.0.2"}, {"v2.0.0", "v2.0.1", "v2.0.2", "v2.0.3", "v2.1.0-beta1", "v2.1.0-beta2"}, {"master", "v2.0", "v2.1"})
v2.1 branches
> semver.select("^2.1", {"2.0.0", "2.0.1", "2.0.2"}, {"v2.0.0", "v2.0.1", "v2.0.2", "v2.0.3", "v2.1.0-beta1", "v2.1.0-beta2"}, {"master", "v2.0", "v2.1"})
/usr/local/share/xmake/core/base/os.lua:592: Unable to select version for range '^2.1'
stack traceback:
        [C]: in function 'error'
        /usr/local/share/xmake/core/base/os.lua:592: in function 'raise'
        /usr/local/share/xmake/core/sandbox/modules/os.lua:423: in function 'raise'
        ...e/xmake/core/sandbox/modules/import/core/base/semver.lua:78: in function <...e/xmake/core/sandbox/modules/import/core/base/semver.lua:73>
        [C]: in function 'interactive'
        ...ake/core/sandbox/modules/import/core/sandbox/sandbox.lua:64: in function 'interactive'
        /usr/local/share/xmake/plugins/lua/xmake.lua:76: in function </usr/local/share/xmake/plugins/lua/xmake.lua:32>
        [C]: in function 'load'
        /usr/local/share/xmake/core/project/task.lua:428: in function 'run'
        /usr/local/share/xmake/core/main.lua:201: in function </usr/local/share/xmake/core/main.lua:159>
```

**Comment by waruqi** [comment]
@uael Great! 👍

**Comment by uael** [comment]
You updated the roadmap, this is all going fine, features are all cool and well designed, good job @waruqi 
When did you think a beta may come ? I will be happy to beta-test

**Comment by waruqi** [comment]
@uael I'm improving the build script on windows and installation, linking, and cross-compilation are not yet implemented. But you can test it first. for example:

```bash
$ cd tests/projects/deps_console_c
$ xmake require 
$ xmake require -v -f
...
```

**Comment by uael** [comment]
@waruqi The command took really long time, did you know why ?

**Comment by waruqi** [comment]
@uael You can add `-v` to see the verbose output. Maybe clone repositories, test host speed, get git tags...  I will optimize performance in the future, but it's not the point now.

**Comment by uael** [comment]
It freeze here `updating repositories .. ` and after
```bash
pulling repository(xmake-repo): https://github.com/tboox/xmake-repo.git to /home/uael/.xmake/repositories/xmake-repo ..
git     pull origin master
From https://github.com/tboox/xmake-repo
 * branch            master     -> FETCH_HEAD
Already up-to-date.
ok
```
for 30s 1m without verbose then it fail on install with:
```bash
  => building tboox.tbox-v1.6.0 .. 
xmake f -p linux -a x86_64 -c
checking for the c include unistd.h ... no
checking for the c include sys/socket.h ... no
checking for the c include locale.h ... no
checking for the c include sys/poll.h ... no
checking for the c include signal.h ... no
checking for the c include sys/uio.h ... no
checking for the c include math.h ... no
checking for the c include wchar.h ... no
checking for the c include unistd.h ... no
checking for the c include unistd.h ... no
checking for the c include ifaddrs.h ... no
checking for the links polarssl ... no
checking for the c include string.h ... no
checking for the c include wchar.h ... no
checking for the c include math.h ... no
checking for the c include string.h ... no
checking for the c include math.h ... no
checking for the c include wchar.h ... no
checking for the c include math.h ... no
checking for the c include string.h ... no
checking for the c include stdlib.h ... no
checking for the c include string.h ... no
checking for the c include wchar.h ... no
checking for the c include sys/sem.h ... no
checking for the c include sys/sendfile.h ... no
checking for the c include string.h ... no
checking for the c include string.h ... no
checking for the c include string.h ... no
checking for the links sqlite3 ... no
checking for the c include sys/uio.h ... no
checking for the c include wchar.h ... no
checking for the c include sys/wait.h ... no
checking for the c include math.h ... no
checking for the c include string.h ... no
checking for the c include time.h ... no
checking for the c include unistd.h ... no
checking for the c include unistd.h ... no
checking for the c include math.h ... no
checking for the c include string.h ... no
checking for the c include math.h ... no
checking for the c include regex.h ... no
checking for the c include math.h ... no
checking for the c include wchar.h ... no
checking for the c include sys/uio.h ... no
checking for the c include math.h ... no
checking for the c type wchar_t ... no
checking for the c include pthread.h ... no
checking for the c include signal.h ... no
checking for the c include string.h ... no
checking for the c include spawn.h ... no
checking for the c include time.h ... no
checking for the c include math.h ... no
checking for the c include execinfo.h ... no
checking for the c include sched.h ... no
checking for the c include math.h ... no
checking for the c include wchar.h ... no
checking for the c include unistd.h ... no
checking for the c include dlfcn.h ... no
checking for the c include regex.h ... no
checking for the c include unistd.h ... no
checking for the c include wchar.h ... no
checking for the c include sys/uio.h ... no
checking for the c include sys/epoll.h ... no
checking for the c include sys/sem.h ... no
checking for the c include pthread.h ... no
checking for the links ssl, crypto ... no
checking for the c include sys/stat.h ... no
checking for the c include pthread.h ... no
checking for the links z ... no
checking for the c include time.h ... no
checking for the c include math.h ... no
checking for the c include signal.h ... no
checking for the c include math.h ... no
checking for the c include string.h ... no
checking for the c include string.h ... no
checking for the c include string.h ... no
checking for the c include pthread.h ... no
checking for the c include string.h ... no
checking for the c include stdio.h ... no
checking for the c include unistd.h ... no
checking for the c include wchar.h ... no
checking for the c include string.h ... no
checking for the c include math.h ... no
checking for the c include semaphore.h ... no
checking for the c include math.h ... no
checking for the c include math.h ... no
checking for the c include unistd.h ... no
checking for the links pcre ... no
checking for the c include sys/time.h ... no
checking for the c include pthread.h ... no
checking for the c include wchar.h ... no
checking for the links pcre2-8 ... no
checking for the c include wchar.h ... no
checking for the links mysqlclient ... no
checking for the c include dirent.h ... no
checking for the c include math.h ... no
checking for the c include string.h ... no
checking for the c include math.h ... no
checking for the c include math.h ... no
checking for the c include wchar.h ... no
checking for the c include sys/epoll.h ... no
checking for the c include string.h ... no
checking for the c include wchar.h ... no
checking for the c include wchar.h ... no
checking for the c include math.h ... no
checking for the links mbedtls, mbedcrypto, mbedx509 ... no
checking for the c include pthread.h ... no
checking for the c include wchar.h ... no
checking for the c include math.h ... no
checking for the c include math.h ... no
checking for the c include signal.h ... no
checking for the c include math.h ... no
checking for the c include unistd.h ... no
checking for the c include math.h ... no
checking for the c include math.h ... no
checking for the c include math.h ... no
checking for the c include sys/resource.h ... no
checking for the c include stdlib.h ... no
checking for the c include unistd.h ... no
checking for the links pthread, dl, m, c ... no
checking for the c include string.h ... no
configure
{
    deprecated = true
,   clean = true
,   polarssl = false
,   object = true
,   arch = x86_64
,   sqlite3 = false
,   info = true
,   smallest = false
,   cc = gcc
,   plat = linux
,   wchar = false
,   float = true
,   host = linux
,   charset = true
,   xml = true
,   exception = false
,   pcre2 = false
,   openssl = false
,   zlib = false
,   database = true
,   regex = true
,   ccache = true
,   buildir = build
,   mysql = false
,   mode = release
,   hash = true
,   mbedtls = false
,   coroutine = true
,   zip = true
,   demo = true
,   base = false
,   kind = static
,   pcre = false
}
xmake -r
error: cannot get tool for dc-ar
/usr/local/share/xmake/core/base/os.lua:592: exec(xmake -r) failed(255)!
stack traceback:
    [C]: in function 'error'
    [/usr/local/share/xmake/core/base/os.lua:592]: in function 'raise'
    [/usr/local/share/xmake/core/sandbox/modules/os.lua:318]:
    [/usr/local/share/xmake/core/sandbox/modules/os.lua:262]: in function 'vrun'
    [/usr/local/share/xmake/actions/require/action/build.lua:43]:
    [/usr/local/share/xmake/actions/require/action/build.lua:100]:
```

**Comment by uael** [comment]
There is a lot more verbose while building i just pasted the end

**Comment by waruqi** [comment]
@uael Ok, I will look at it.

**Comment by waruqi** [comment]
@uael You can try running `xmake require -v -f` to force to download package again and give me verbose info. 

About this problem, you can create a new issue. 😸

**Comment by uael** [comment]
I will :+1:

**Comment by waruqi** [comment]
@uael I improved the speed of pulling git refs (multi-jobs) and cache it, you can try it again. But it is still slower in the first time. I will continue to improve it in the future.

**Comment by uael** [comment]
I just updated from the latest repo
```bash
uael@aero:~/xmake/tests/projects/deps_console_c$ xmake require -v -f
error: /home/uael/xmake/tests/projects/deps_console_c/xmake.lua:2: attempt to call global 'package' (a nil value)
stack traceback:
    [/home/uael/xmake/tests/projects/deps_console_c/xmake.lua:2]: in main chunk
```
It's maybe badly installed from me :smile_cat: 
I run make build then the get.sh with `__install_only__` and `__local__`

**Comment by waruqi** [comment]
@uael This should be a installation problem. It will be installed to `/usr/local` directory by default if we run `make build; sudo make install`. And it will be installed to `~/.local` directory if we run `scripts/get.sh __local__`.

So maybe the previous version has not been uninstalled or updated and you're running the old version of xmake.

You can attempt to run `sudo make uninstall` to remove old version first and try installing it again.

**Comment by uael** [comment]
Hi @waruqi, how is this going ? I'm currently passing some exams cannot follow as before 😢

**Comment by waruqi** [comment]
@uael Thank you very much for your contribution and help. The semantic version module is very great now. 

I'm implementing `find_package` to detect and link the dependent packages and I'll finish the rest of the job as soon as possible. 

And I haven't had much time to focus on semver module lately. If there is a problem with the semantic version in the future, I will try to modify it and submit pr and issues to you.  

Thanks!

**Comment by waruqi** [comment]
@uael  I'm so sorry I spent too much time in order to implement the detect modules, so the development of package management was delayed. 

Now I'm going to continue to start this job and hope to be able to complete it in v2.2.1 version.

I have merged all patches of the `repo` branch to `dev` branch and will remove `repo` branch.

Because there are too many conflicts, I can only merge code manually, and I'm sorry that I didn't keep the contributors ' information. 😭

**Comment by waruqi** [comment]
A test example:

```console
$ cd xmake/tests/project/requires
$ xmake
```

I am improving the dependent environment of requires on each platform (git, curl, ..).

**Comment by yangjz1125** [comment]
Can I usr remote package management now?

**Comment by waruqi** [comment]
@yangjz1125 Not yet, please wait some time.

**Comment by yangjz1125** [comment]
Okay, thanks.

**Comment by waruqi** [comment]
An example: https://github.com/tboox/benchbox

Run:

```console
$ git clone https://github.com/tboox/benchbox.git
$ cd benchbox
$ xmake
```

![package](https://user-images.githubusercontent.com/151335/45801222-d57be780-bce5-11e8-8903-b21e1ad6b22d.gif)

**Comment by waruqi** [comment]
@yangjz1125 @uael This feature can be used for v2.2.2

An example project: https://github.com/tboox/benchbox

Documents: 

* [English](https://xmake.io/#/home?id=remote-dependency-mode)
* [中文](https://xmake.io/#/zh/?id=%E8%BF%9C%E7%A8%8B%E4%BE%9D%E8%B5%96%E6%A8%A1%E5%BC%8F)

```lua
add_requires("tbox 1.6.*", "libpng ~1.16", "zlib")

target("test")
    set_kind("binary")
    add_files("src/*.c") 
    add_packages("tbox", "libpng", "zlib")
```

<img src="https://xmake.io/assets/img/index/package_manage.png" width="80%" />

<img src="https://xmake.io/assets/img/index/package_arch.png" width="80%" />

**Comment by uael** [comment]
Very good work here, repo/package are a well designed approach. I really have to pick a side project to try it out!

**Comment by uael** [comment]
I just finished the docs, thank you for quoting me 😃 Very accurate and understandable by the way

**Comment by waruqi** [comment]
I have almost completed this feature, and I will continue to improve it and the package repository. [xmake-repo](https://github.com/xmake-io/xmake-repo)

Documents: https://xmake.io/#/home?id=remote-dependency-mode
